### DISCUSSION TITLE
How to handle reusable routes and keep typesafety?

### DISCUSSION THREAD

**Discussion by SteNafH** [TanStack/router#1108] [General]
Hello everyone, I'm currently developing an application that involves checking dependencies for repositories. I recently
came across `@tanstack/router` and was impressed by its type-safe routing and the way it handles search parameters.

However, I'm facing challenges in implementing routing for my application, particularly when dealing with reusable
routes. For instance, my "Dependency" route should be accessible through multiple paths. I'm wondering if there are any
examples available on how to effectively work with reusable routes.

I'm uncertain whether I need to manually specify each route. My concern is that if I do this, I might lose
the benefits of type safety, as the `{from: ...}` needs to be specified in both `useSearch` and `useParams`. Are there
any suggestions or examples on how to maintain type safety while working with reusable routes in this context?

Underneath is hopefully a more detailed visualization of my application routes:

```
- /repositories (Lists all repositories)
  - /:repository (Lists commits for repository)
    - /:commit (Lists dependencies for commit)
      - /:dependency (Lists cves for dependency)

- /dependencies (Lists all dependencies)
  - /:dependency (Lists commits that use dependency)
    - /cves (Lists cves for dependency)
    - /:commit (Lists dependencies for commit)
      - /:dependency (Lists cves for dependency)

- /licenses (Lists all licenses)
  - /:license (Lists all dependencies that use license)
    - /:dependency (Lists all commits that use dependency)
      - /cves (Lists cves for dependency)
      - /:commit (Lists dependencies for commit)
        - /:dependency (Lists cves for dependency)
```

A graph for the application flow:

```
Dependencies   Repositories        Licenses
     \               |               /
   Dependency   Repository       License
       \             |             /
        \ ──────> Commit  <── Dependency
         \           |           /  
          \──────> Cves <───────/
```

I prefer to youse code based routing, but any insight is welcome!
Thanks in advance

**Comment by thyming** [comment]
my workaround for this has been to do the following:
* export the `typeof Route` for each route you're going to be reusing a component for
* wrap the component with a provider that provides the route object to the reused component
* use the route from the context in the component, which will have the route types intersected in a way that it will be type-safe

for example:
Ctx.tsx
```typescript
type Routes = typeof Route1 | typeof Route2
export const Ctx = React.createContext({} as Routes)
```
Route1.tsx
```typescript
export const Route1 = createFileRoute('/route1')({
  component: () => <Ctx.Provider value={Route1}><ReusedComponent /></Ctx.Provider>,
  loader: () => Promise.resolve("1")
})
```
Route2.tsx
```typescript
export const Route2 = createFileRoute('/route2')({
  component: () => <Ctx.Provider value={Route2}><ReusedComponent /></Ctx.Provider>,
  loader: () => Promise.resolve("2")
})
```
ReusedComponent.tsx
```typescript
export const ReusedComponent = () => {
  const route = useContext(Ctx);
  const data = route.useLoaderData(); // this will be correctly typed as string
  return <pre>{JSON.stringify(data)}</pre>
}
```
This is pretty clunky, but it does achieve the objective of sharing route context in a component across multiple routes. Maybe someone has an idea about how to support this functionality in a bit more elegant way?

**Comment by trabati** [comment]
A year has passed, is there a better way today?

**Comment by ferretwithaberet** [comment]
I am curious if anyone found a better way to reuse entire route options. I want to define generic logic that will be mounted on multiple routes. Most of the time I only need to reuse `beforeLoad`, `loader` or `component`, but it'd be nice to be able to reuse the entire object.

**Comment by julen** [comment]
Hey folks, I'd like to revive this discussion as we are facing a similar challenge with TanStack Start.

Our use case: we are organizing our codebase by domain modules rather than URL structure. For example, our `src/videos/routes.ts` module needs to be mountable at multiple paths: `/recordings`, `/org/$orgId/recordings`, etc. This is a common pattern when building modular applications where domain logic doesn't map 1:1 with URLs.

### What we've tried:

* **Virtual file routes**: while these provide great flexibility, the convention requires each referenced module to export exactly one `Route` symbol (per the implementation in [packages/router-generator/src/transform/transform.ts](https://github.com/TanStack/router/blob/d6f5870122c69829ed55f296f5b0f306465f2887/packages/router-generator/src/transform/transform.ts#L111-L154)). This means we can't directly mount the same route module at multiple paths.
* **Creating duplicate route files**: as suggested in #5009, we could create separate files that re-export or redefine routes. However, with dozens of routes per module, this approach isn't practical or scalable for our team.
* **Code-based routing**: we thought this could help as we'd be creating the route tree ourselves, but it is unsupported in TanStack Start (#5599).

### Help needed

We are looking for guidance on how to handle this scenario. Route mounting/inclusion has been a standard feature in web frameworks for decades (e.g., [Django's `include()`](https://docs.djangoproject.com/en/5.2/topics/http/urls/#including-other-urlconfs)), so we are hoping there's either:

* A pattern we are missing that enables this today.
* Openness to considering enhancements that would support mounting route modules at multiple paths.

We understand TSR's focus on type safety, and we value that. But the ability to compose and reuse route modules feels foundational for building maintainable applications.

@schiller-manuel would love your perspective on this.

Thank you!
